crates/aquascope/src/analysis/permissions/flow.rs

```diff
@@ -0,0 +1,474 @@
+  }
+
+  pub fn has_local_member(&self, origin: Origin) -> bool {
+    self.contains_local.count(self.scc(origin)) > 0
+  }
+
+  /// Get the specific kind of flow edge that connects `from` and `to`.
+  #[allow(clippy::match_same_arms)]
+  pub(crate) fn flow_kind(&self, from: Origin, to: Origin) -> FlowEdgeKind {
+    let scc_from = self.constraint_graph.scc(from);
+    let scc_to = self.constraint_graph.scc(to);
+
+    // Data can always flow within the same SCC.
+    if scc_from == scc_to {
+      return FlowEdgeKind::Ok;
+    }
+
+    let from_contains_abstract = self.contains_abstract.count(scc_from) > 0;
+    let from_contains_local = self.contains_local.count(scc_from) > 0;
+
+    let to_contains_abstract = self.contains_abstract.count(scc_to) > 0;
+    let to_contains_local = self.contains_local.count(scc_to) > 0;
+
+    log::debug!(
+      "Analyzing flow {from:?} -> {to:?} {:?} -> {:?}",
+      self.scc(from),
+      self.scc(to)
+    );
+
+    log::debug!("{from:?} is_local? {from_contains_local} is_abstract? {from_contains_abstract}", );
+    log::debug!("{to:?} is_local? {to_contains_local} is_abstract? {to_contains_abstract}", );
+
+    // A local value can never flow into a concrete region.
+    if from_contains_local && to_contains_abstract {
+      log::debug!("early return local outlives universal");
+      return FlowEdgeKind::LocalOutlivesUniversal;
+    }
+
+    // If both regions contain abstract, we check that all regions in `from`
+    // are known to outlive those in `to`. Otherwise, there would be a missing
+    // constraint.
+    if !self.contains_abstract.iter(scc_from).all(|from| {
+      self
+        .contains_abstract
+        .iter(scc_to)
+        .all(|to| self.known_flows.contains(from, to))
+    }) {
+      return FlowEdgeKind::MissingUniversalConstraint;
+    }
+
+    // If `from` is flowing a dangling pointer we would always consider this an error.
+    if self
+      .contains_local
+      .iter(scc_from)
+      .any(|local| self.dangling_local_sources.contains(local))
+    {
+      return FlowEdgeKind::LocalInvalidatedAtExit;
+    }
+
+    FlowEdgeKind::Ok
+  }
+}
+
+// ------------------
+// Utilities
+
+fn flatten_tuples<T>(tups: &[(T, T)]) -> impl Iterator<Item = T> + '_
+where
+  T: Copy + Clone + Eq + std::hash::Hash,
+{
+  tups.iter().flat_map(|&(o1, o2)| [o1, o2]).unique()
+}
+
+fn count_nodes<T: Idx>(tups: &[(T, T)]) -> usize {
+  flatten_tuples(tups)
+    .minmax_by_key(|v| v.index())
+    .into_option()
+    .map_or(0, |(_, mx)| mx.index() + 1)
+}
+
+/// Compute the transitive flows from a set of given `sources` in `graph`.
+fn flow_from_sources<T>(
+  mat: &mut BitMatrix<T, T>,
+  sources: impl Iterator<Item = T>,
+  graph: impl DirectedGraph<Node = T> + WithSuccessors + WithNumNodes,
+) where
+  T: Idx,
+{
+  // Initialize all abstract sources.
+  for a in sources {
+    mat.insert(a, a);
+  }
+
+  let mut changed = true;
+
+  while changed {
```

Comment:
Can we use the [`TransitiveRelation`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/transitive_relation/struct.TransitiveRelation.html) data structure instead? I think that's already used in rustc for reachability queries.

Reply:
Indeed we can! I didn't know this existed :)